File: C_plus_plus/Flatten_Binary_Tree_to_Linked_List.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode *root) {
        insert(root);
    }
    TreeNode * insert(TreeNode *root){
        TreeNode *last;
        if(!root) return NULL;
        if(!root->left&&!root->right) return root;
        TreeNode * right = root->right;
        if(root->left){
            root->right = root->left;
            root->left=NULL;
            last = insert(root->right);
            if(right){
                last->right = right;
                return insert(right);
            }
            else return last;
        }
        else return insert(right);
    }
};

File: java/Flatten_Binary_Tree_to_Linked_List.java
// Recursive solution, DFS
// Move the left tree to the root's right , and the right tree to the left's tail.
// Time is O(n), space is O(n).

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void flatten(TreeNode root) {
        generate(root);
    }
    public Pack generate(TreeNode root) {
        if (root == null) return new Pack(null,null);
        if (root.left == null && root.right == null)
            return new Pack(root,root);
        Pack p1 = generate(root.left);
        Pack p2 = generate(root.right);
        root.left = null;
        if (p1.head == null) {
            root.right = p2.head;
            p2.tail.right = null;
            return new Pack(root,p2.tail);
        }
        else if (p2.head == null) {
            root.right = p1.head;
            p1.tail.right = null;
            return new Pack(root,p1.tail);
        }
        else {
            root.right = p1.head;
            p1.tail.right = p2.head;
            p2.tail.right = null;
            return new Pack(root,p2.tail);
        }
    }
    class Pack {
        TreeNode head, tail;
        public Pack(TreeNode head, TreeNode tail) {
            this.head = head;
            this.tail = tail;
        }
    }
}
